### PR TITLE
Allow users to connect to their GitHub accounts

### LINKED ISSUE #1894: Allow users to connect their accounts to Github
See v1:
- [buildOauthConsentUrl](https://github.com/OpenFn/platform-app/blob/main/assets/js/components/UserEdit.js#L80)
- https://github.com/OpenFn/platform-app/blob/main/assets/js/helpers/index.js#L21
- Connect to github [figma](https://www.figma.com/file/2Uvr2jOaQh6rQNR1cHlLKM/App?type=design&node-id=2086%3A2303&mode=design&t=OEZgJslyfupUtLUE-1)
- Remove from github [figma](https://www.figma.com/file/2Uvr2jOaQh6rQNR1cHlLKM/App?type=design&node-id=2090%3A2409&mode=design&t=OEZgJslyfupUtLUE-1)

User Settings (button color wrong here)

Image

Image

### PR DESCRIPTION
**Pull request by midigofrank**
### Demo:
https://github.com/OpenFn/lightning/assets/39288959/caa3f05f-49c3-4f1a-9a41-1f4573383def
### Required ENVS:
- `GITHUB_APP_CLIENT_ID`
- `GITHUB_APP_CLIENT_SECRET`
### Checklist

- [x] User can connect
- [x] User can disconnect 

## Notes for the reviewer

- Adds a new column, `github_oauth_token` to the `users` table. This field is encrypted at rest, with the help of `Lightning.Encrypted.Map`
- Adds a new route, `/oauth/:provider/callback` and controller `OauthController`. I've decided NOT to use the existing route and controller `OidcController` because this new logic deals with users who are logged in. Maybe I should merge them in together? 🤷‍♂️ 
- Github supports both non-expiry and expiry tokens. I have handled both cases
- I have defined a global test adapter for `tesla`, `config :tesla, adapter: Lightning.Tesla.Mock`. I think I should have done this in https://github.com/OpenFn/lightning/pull/1781 instead of defining only one for the `GithubClient`. Consequently, I have been forced to `stub` this with the `Hackney adapter` so that the `Bypass` tests work. 

## Related issue

Fixes #1894 

## Review checklist

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** (`:owner`, `:admin`, `:editor`, `:viewer`) have been implemented and tested. 
- [x] If needed, I have updated the **changelog**
- [x] Product has **QA'd** this feature
- [x] The callback urls have been updated in the `prod` and `demo` Github Apps Setting

### CODE REVIEW

**Review by taylordowns2000 — APPROVED** [review]

File: config/runtime.exs [hunk 1]
```diff
@@ -62,12 +62,24 @@ github_app_id =
...
+
+github_app_client_id =
+  System.get_env("GITHUB_APP_CLIENT_ID") ||
+    Application.get_env(:lightning, :github_app, [])
+    |> Keyword.get(:client_id, nil)
+
+github_app_client_secret =
+  System.get_env("GITHUB_APP_CLIENT_SECRET") ||
+    Application.get_env(:lightning, :github_app, [])
+    |> Keyword.get(:client_secret, nil)
```
Comment (jyeshe) [inline config/runtime.exs:75]: If mandatory on prod, would move to prod section and use a `Keyword.fetch!`. Unless we have some instances where the Github feature would be turned off (possibly on a cluster or set of nodes for free users). cc @taylordowns2000

File: lib/lightning/version_control/version_control.ex [hunk 2]
```diff
@@ -123,6 +127,197 @@ defmodule Lightning.VersionControl do
...
+        false
+    end
+  end
+
+  @doc """
+  Fecthes the access token for the given `User`.
+  If the access token has expired, it `refreshes` the token and updates the `User` column accordingly
+  """
+  @spec fetch_user_access_token(User.t()) ::
+          {:ok, String.t()} | {:error, :expired | map()}
```
Comment (jyeshe) [inline lib/lightning/version_control/version_control.ex:216]: The spec is a bit different `{:error, :expired}`. Could you double check the error handling?

File: lib/lightning/version_control/version_control.ex [hunk 3]
```diff
@@ -8,11 +8,15 @@ defmodule Lightning.VersionControl do
...
   import Ecto.Query, warn: false
 
+  alias Lightning.Accounts.User
   alias Lightning.Repo
+  alias Lightning.VersionControl.Events
   alias Lightning.VersionControl.GithubClient
   alias Lightning.VersionControl.GithubError
   alias Lightning.VersionControl.ProjectRepoConnection
 
+  defdelegate subscribe(user), to: Events
```
Comment (jyeshe) [inline lib/lightning/version_control/version_control.ex:18]: :+1:

**Review by jyeshe — COMMENTED** [review]
@midigofrank this is great. Found some opportunities for fail fast, to know directly the root cause of some errors and there is one drift on spec for error handling. Wondering why dialyzer has not caught it :thinking:

File: config/runtime.exs (hunk 1 above)
Comment (jyeshe) [reply config/runtime.exs:75]: I have talked with Taylor and the code is correct. The feature might not be used on prod on certain instances.

**Review by jyeshe — CHANGES_REQUESTED** [review]
Hey @midigofrank I wasn't clear about the timing for this changes (considering the latest planning). I would appreciate if you had a look specially at the error spec and on the fail fast when `client_id` is `nil`. These are important to not affect our MTTR in the future. The other one is a quick/simple style thing. Thank you!

**Review by jyeshe — APPROVED** [review]
This is great! 🌟

PS: The Github disconnect is on a parallel thread on slack (so ignoring the CI error for this PR)

**Pull request merged**